### PR TITLE
[Serialization] Deserialize custom attributes after the attached decl

### PR DESCRIPTION
**Pull request by xymus**
With type-wrappers a custom attribute may reference a type and lead to a cycle in deserialization if the target type references the type-wrapper. To avoid this scenario, move deserializing the custom decls from before to after the decl they are attached to.

rdar://103425758

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
[swift-ci](https://github.com/swift-ci) Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: test/IDE/print_ast_tc_decls.swift [hunk 1]
```diff
@@ -547,7 +547,10 @@ class d0170_TestAvailability {
 // PASS_COMMON-LABEL: {{^}}@objc class d0180_TestIBAttrs {{{$}}
 
   @IBAction func anAction(_: AnyObject) {}
-// PASS_COMMON-NEXT: {{^}}  @objc @IBAction @MainActor func anAction(_: AnyObject){{$}}
+// PASS_COMMON-NEXT: {{^}}  @objc
+// PASS_COMMON-DAG: @IBAction
+// PASS_COMMON-DAG: @MainActor
+// PASS_COMMON: func anAction(_: AnyObject){{$}}
```
[the remark this answers is not in the document]
Comment (xymus) [reply test/IDE/print_ast_tc_decls.swift:553]: We get a different attribute order between line 13 and line 35. That test is cryptic to me but as I understand it, one reads directly from the source and the latter one from a compiled swiftmodule file. The one reading from the swiftmodule file has the change in the ordering.

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: lib/Serialization/Deserialization.cpp [hunk 2]
```diff
@@ -2740,6 +2740,7 @@ class DeclDeserializer {
 
   DeclAttribute *DAttrs = nullptr;
   DeclAttribute **AttrsNext = &DAttrs;
+  llvm::SmallVector<serialization::BitOffset> customAttrOffsets;
```
[the remark this answers is not in the document]
Comment (xymus) [reply lib/Serialization/Deserialization.cpp]: Done.

File: lib/Serialization/Deserialization.cpp [hunk 3]
```diff
@@ -2828,8 +2829,15 @@ class DeclDeserializer {
 
   /// Deserializes records common to all decls from \c MF.DeclTypesCursor (ie.
   /// the invalid flag, attributes, and discriminators)
+  ///
+  /// Skips the custom attributes and append their offsets to
```
[the remark this answers is not in the document]
Comment (xymus) [reply lib/Serialization/Deserialization.cpp]: I clarified the comment a bit.

File: lib/Serialization/Deserialization.cpp [hunk 4]
```diff
@@ -5074,33 +5136,17 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
...
+        // Deserialize the custom attributes after the attached decl,
+        // skip for now.
+        customAttrOffsets.push_back(attrOffset);
+        skipAttr = true;
+
+        // Skip the record.
         bool isImplicit;
         bool isArgUnsafe;
         TypeID typeID;
         serialization::decls_block::CustomDeclAttrLayout::readRecord(
```
[the remark this answers is not in the document]
Comment (xymus) [reply lib/Serialization/Deserialization.cpp]: Right, we can remove it.

File: lib/Serialization/Deserialization.cpp [hunk 5]
```diff
@@ -5386,19 +5432,24 @@ DeclDeserializer::getDeclCheckedImpl(
...
       setOriginalDeclarationAndParameterIndicesInDifferentiableAttributes(\
-          decl.get(), DAttrs, diffAttrParamIndicesMap); \
+          declOrError.get(), DAttrs, diffAttrParamIndicesMap); \
     } \
-    return decl; \
+    if (declOrError) \
+      result = declOrError.get(); \
+    else \
+      return declOrError; \
+    break; \
```
[the remark this answers is not in the document]
Comment (xymus) [reply lib/Serialization/Deserialization.cpp]: Yeah, might as well.

File: lib/Serialization/Deserialization.cpp [hunk 6]
```diff
@@ -5432,15 +5483,24 @@ DeclDeserializer::getDeclCheckedImpl(
...
     auto resolved = MF.resolveCrossReference(baseModuleID, pathLen);
-    if (resolved)
+    if (resolved) {
       declOrOffset = resolved.get();
-    return resolved;
+      result = resolved.get();
+    } else
+      return resolved;
+    result = resolved.get();
+    break;
```
[the remark this answers is not in the document]
Comment (xymus) [reply lib/Serialization/Deserialization.cpp]: We can do without `result` so it simplifies the code quite a bit.

File: test/IDE/print_ast_tc_decls.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (xymus) [reply test/IDE/print_ast_tc_decls.swift]: I added a comment explaining the different expected behaviors.

**Review by xedin — APPROVED** [review]

File: lib/Serialization/Deserialization.cpp [hunk 7]
```diff
@@ -5389,16 +5428,19 @@ DeclDeserializer::getDeclCheckedImpl(
...
       // attributes. \
       */ \
       setOriginalDeclarationAndParameterIndicesInDifferentiableAttributes(\
-          decl.get(), DAttrs, diffAttrParamIndicesMap); \
+          declOrError.get(), DAttrs, diffAttrParamIndicesMap); \
     } \
-    return decl; \
+    if (!declOrError) \
+      return declOrError; \
+    declOrOffset = declOrError.get(); \
```
Comment (bnbarham) [inline lib/Serialization/Deserialization.cpp:5442]: I think each `deserialize*` method already does this 😅. Suppose it doesn't hurt though.

**Review by bnbarham — APPROVED** [review]

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Pull request merged**